internal/assetloader/loader.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2025 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package assetloader
+
+import (
+	"errors"
+	"os/exec"
+	"slices"
+
+	"github.com/palantir/godel-conjure-plugin/v6/internal/assetapi"
+	"github.com/palantir/godel-conjure-plugin/v6/internal/cmdutils"
+	"github.com/palantir/godel-conjure-plugin/v6/internal/conjureextensionsprovider"
+	pkgerrors "github.com/pkg/errors"
+)
+
+type LoadedAssets struct {
+	ConjureExtensionsProviders []conjureextensionsprovider.Generator
+
+	ConjureIRExtensionsProviders []string
+}
+
+// LoadAssets takes a list of asset paths and returns a LoadedAssets struct that contains the typed assets. Returns an
+// error if any of the provided assets are not valid according to the plugin asset specification.
+func LoadAssets(assets []string) (LoadedAssets, error) {
+	assetTypeToAssetsMap, err := createAssetTypeToAssetsMap(assets)
+	if err != nil {
+		return LoadedAssets{}, err
+	}
+
+	// create Conjure extensions generators from assetapi.ConjureExtensionsProvider assets
+	var conjureExtensionsProviders []conjureextensionsprovider.Generator
+	for _, assetPath := range assetTypeToAssetsMap[assetapi.ConjureExtensionsProvider] {
+		conjureExtensionsProviders = append(conjureExtensionsProviders, conjureextensionsprovider.NewAssetProvider(assetPath))
+	}
+
```

Comment:
Constraints like "only 1 asset of backcompat type" could be checked for/enforced in this function